### PR TITLE
[4.3] Use the Web Asset Manager

### PR DESCRIPTION
**Pull request by brianteeman**
As all the htmlhelper behviors are deprecated this PR continues the work to remove their usage in core.

We should not be using deprecated code.

### CODE REVIEW

File: administrator/components/com_redirect/layouts/toolbar/batch.php [hunk 1]
```diff
@@ -12,7 +12,9 @@
 
 use Joomla\CMS\HTML\HTMLHelper;
```
Comment (Quy) [inline administrator/components/com_redirect/layouts/toolbar/batch.php]: Please remove.

File: administrator/components/com_categories/tmpl/categories/modal.php [hunk 2]
```diff
@@ -25,7 +25,9 @@
     Session::checkToken('get') or die(Text::_('JINVALID_TOKEN'));
 }
 
-HTMLHelper::_('behavior.core');
+/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
+$wa = $app->getDocument()->getWebAssetManager();
```
Comment (HLeithner) [inline administrator/components/com_categories/tmpl/categories/modal.php]:
```suggestion
$wa = $this->document->getWebAssetManager();
```

File: administrator/components/com_messages/tmpl/message/default.php [hunk 3]
```diff
@@ -14,7 +14,10 @@
 use Joomla\CMS\Language\Text;
 use Joomla\CMS\Router\Route;
 
-HTMLHelper::_('behavior.core');
+/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
+$wa = $app->getDocument()->getWebAssetManager();
```
Comment (HLeithner) [inline administrator/components/com_messages/tmpl/message/default.php]:
```suggestion
$wa = $this->document->getWebAssetManager();
```

File: administrator/components/com_newsfeeds/tmpl/newsfeeds/modal.php [hunk 4]
```diff
@@ -18,7 +18,9 @@
 use Joomla\CMS\Router\Route;
 use Joomla\Component\Newsfeeds\Site\Helper\RouteHelper;
 
-HTMLHelper::_('behavior.core');
+/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
+$wa = $app->getDocument()->getWebAssetManager();
```
Comment (HLeithner) [inline administrator/components/com_newsfeeds/tmpl/newsfeeds/modal.php]:
```suggestion
$wa = $this->document->getWebAssetManager();
```

File: administrator/components/com_redirect/layouts/toolbar/batch.php [hunk 5]
```diff
@@ -10,9 +10,9 @@
 
 defined('_JEXEC') or die;
 
-use Joomla\CMS\HTML\HTMLHelper;
-
-HTMLHelper::_('behavior.core');
+/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
+$wa = $app->getDocument()->getWebAssetManager();
```
Comment (HLeithner) [inline administrator/components/com_redirect/layouts/toolbar/batch.php]:
```suggestion
$wa = \Joomla\CMS\Factory\Factory::getApplication()->getDocument()->getWebAssetManager();
```

File: layouts/joomla/sidebars/submenu.php [hunk 6]
```diff
@@ -10,11 +10,13 @@
 
 defined('_JEXEC') or die;
 
-use Joomla\CMS\HTML\HTMLHelper;
 use Joomla\CMS\Language\Text;
 use Joomla\Filter\OutputFilter;
 
-HTMLHelper::_('behavior.core');
+/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
+$wa = $app->getDocument()->getWebAssetManager();
```
Comment (HLeithner) [inline layouts/joomla/sidebars/submenu.php]:
```suggestion
$wa = \Joomla\CMS\Factory\Factory::getApplication()->getDocument()->getWebAssetManager();
```

File: layouts/joomla/toolbar/batch.php [hunk 7]
```diff
@@ -10,12 +10,13 @@
...
 defined('_JEXEC') or die;
 
-use Joomla\CMS\HTML\HTMLHelper;
 use Joomla\CMS\Language\Text;
 
 // @todo: Deprecate this file since we can use popup button to raise batch modal.
 
-HTMLHelper::_('behavior.core');
+/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
+$wa = $app->getDocument()->getWebAssetManager();
```
Comment (HLeithner) [inline layouts/joomla/toolbar/batch.php]:
```suggestion
$wa = \Joomla\CMS\Factory\Factory::getApplication()->getDocument()->getWebAssetManager();
```

File: modules/mod_login/tmpl/default_logout.php [hunk 8]
```diff
@@ -14,7 +14,10 @@
 use Joomla\CMS\Language\Text;
 use Joomla\CMS\Router\Route;
 
-HTMLHelper::_('behavior.keepalive');
+/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
+$wa = $this->document->getWebAssetManager();
```
Comment (HLeithner) [inline modules/mod_login/tmpl/default_logout.php]:
```suggestion
$wa = \Joomla\CMS\Factory\Factory::getApplication()->getDocument()->getWebAssetManager();
```

**Review by HLeithner — CHANGES_REQUESTED** [review]
* layouts don't have a `$app`
* tmpl files should use the local supplied `document` 
* modules are not in a `Object` scope

**Comment by brianteeman** [comment]
@HLeithner I have merged your suggestions - thanks

I should point out that many existing modules and layouts will need correcting

**Comment by HLeithner** [comment]
> @HLeithner I have merged your suggestions - thanks
> 
> I should point out that many existing modules and layouts will need correcting

possible. for layouts we should change LayoutFile object to supply $document at some point in time. It's not set yet if we like to provide $document or as meta object holding $document and maybe other object like webassetmanager directly.

but for the moment we load the document from the factory

**Comment by brianteeman** [comment]
@hleithner I believe I have resolved all your requested changes. Please can you mark them as complete

**Review by HLeithner — APPROVED** [review]
thanks

File: modules/mod_login/tmpl/default_logout.php [hunk 9]
```diff
@@ -14,7 +14,10 @@
 use Joomla\CMS\Language\Text;
 use Joomla\CMS\Router\Route;
 
-HTMLHelper::_('behavior.keepalive');
+/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
+$wa = \Joomla\CMS\Factory\Factory::getApplication()->getDocument()->getWebAssetManager();
```
Comment (Quy) [inline modules/mod_login/tmpl/default_logout.php]:
```suggestion
$wa = \Joomla\CMS\Factory::getApplication()->getDocument()->getWebAssetManager();
```

File: layouts/joomla/toolbar/batch.php [hunk 10]
```diff
@@ -10,12 +10,13 @@
...
 defined('_JEXEC') or die;
 
-use Joomla\CMS\HTML\HTMLHelper;
 use Joomla\CMS\Language\Text;
 
 // @todo: Deprecate this file since we can use popup button to raise batch modal.
 
-HTMLHelper::_('behavior.core');
+/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
+$wa = \Joomla\CMS\Factory\Factory::getApplication()->getDocument()->getWebAssetManager();
```
Comment (Quy) [inline layouts/joomla/toolbar/batch.php]:
```suggestion
$wa = \Joomla\CMS\Factory::getApplication()->getDocument()->getWebAssetManager();
```

File: layouts/joomla/sidebars/submenu.php [hunk 11]
```diff
@@ -10,11 +10,13 @@
 
 defined('_JEXEC') or die;
 
-use Joomla\CMS\HTML\HTMLHelper;
 use Joomla\CMS\Language\Text;
 use Joomla\Filter\OutputFilter;
 
-HTMLHelper::_('behavior.core');
+/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
+$wa = \Joomla\CMS\Factory\Factory::getApplication()->getDocument()->getWebAssetManager();
```
Comment (Quy) [inline layouts/joomla/sidebars/submenu.php]:
```suggestion
$wa = \Joomla\CMS\Factory::getApplication()->getDocument()->getWebAssetManager();
```

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on ebe6f9424c5c32d89bf9238cbb74271cdfc91880<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38328">issues.joomla.org/tracker/joomla-cms/38328</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on ebe6f9424c5c32d89bf9238cbb74271cdfc91880<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38328">issues.joomla.org/tracker/joomla-cms/38328</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38328">issues.joomla.org/tracker/joomla-cms/38328</a>.</sub>

**Comment by obuisard** [comment]
Thank you Brian @brianteeman for the cleanup!

**Comment by brianteeman** [comment]
thx

**Pull request merged**